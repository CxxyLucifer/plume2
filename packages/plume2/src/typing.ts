import { Map } from 'immutable';
import { ViewAction } from './view-action';

/**
 * Immutable的map类型
 */
export type IMap = Map<string, any>;

/**
 * actor的路由类型
 */
export type TRoute = {
  [name: string]: (state: IMap, params?: any) => IMap;
};

/**
 * Relax容器组件的上下文类型
 */
export interface IRelaxContext<Store> {
  _plume$Store: Store;
}

/**
 * Relax容器组件的props类型
 */
export interface IRelaxProps {
  relaxProps?: Object;
  [name: string]: any;
}

/**
 * Relax容器组件的类型
 */
export interface IRelaxComponent extends React.ComponentClass<IRelaxProps> {
  relaxProps?: Object;
}

/**
 * Store数据容器的参数配置类型
 */
export interface IOptions {
  debug?: boolean;
  [name: string]: any;
}

export interface IViewActionMapper {
  [name: string]: ViewAction;
}

export type TViewAction<T> = {
  [K in keyof T]: T[K] extends new (...args: Array<any>) => infer R ? R : any
};
