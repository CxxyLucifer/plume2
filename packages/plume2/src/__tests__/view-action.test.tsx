import React from 'react';
import renderer from 'react-test-renderer';
import {
  Action,
  Actor,
  IMap,
  Relax,
  Store,
  StoreProvider,
  ViewAction
} from '../index';
import { TViewAction } from '../typing';

// ===========build ViewAction============
class HelloViewAction extends ViewAction {
  sayHello = () => {
    this.store.dispatch('hello', 'hello');
  };
}

class WorldViewAction extends ViewAction {
  sayWorld = () => {
    this.store.dispatch('world', 'world');
  };
}

//===========export viewAction=============
const viewAction = {
  HelloViewAction,
  WorldViewAction
};

//==============build Actor==============
class HelloWorldActor extends Actor {
  defaultState() {
    return {
      hello: '',
      world: ''
    };
  }

  @Action('hello')
  hello(state, msg: string) {
    return state.set('hello', msg);
  }

  @Action('world')
  world(state: IMap, msg: string) {
    return state.set('world', msg);
  }
}

//===============AppStore==================
class AppStore extends Store<typeof viewAction> {
  bindActor() {
    return [new HelloWorldActor()];
  }

  bindViewAction(): any {
    return viewAction;
  }
}

@Relax
class HelloRelax extends React.Component {
  props: {
    relaxProps?: {
      hello: string;
      world: string;
      viewAction: TViewAction<typeof viewAction>;
    };
  };

  static relaxProps = {
    hello: 'hello',
    world: 'world',
    viewAction: 'viewAction'
  };

  render() {
    const { hello, world, viewAction } = this.props.relaxProps;
    expect(Object.keys(viewAction)).toEqual([
      'HelloViewAction',
      'WorldViewAction'
    ]);

    return (
      <div>
        <div>{hello}</div>
        <div>{world}</div>
      </div>
    );
  }
}

@StoreProvider(AppStore, { debug: true })
class HelloApp extends React.Component {
  render() {
    return <HelloRelax />;
  }
}

describe('viewAction test suite', () => {
  it('test viewAction', () => {
    const store = new AppStore();
    store.viewAction.HelloViewAction.sayHello();
    expect(store.get('hello')).toEqual('hello');
    store.viewAction.WorldViewAction.sayWorld();
    expect(store.get('world')).toEqual('world');
  });

  it('test relax inject viewAction', () => {
    const component = renderer.create(<HelloApp />);
    expect(component.toJSON()).toMatchSnapshot();
    const store = window['_plume2App'].HelloApp.store as AppStore;
    store.viewAction.HelloViewAction.sayHello();
    store.viewAction.WorldViewAction.sayWorld();
    expect(component.toJSON()).toMatchSnapshot();
  });
});
