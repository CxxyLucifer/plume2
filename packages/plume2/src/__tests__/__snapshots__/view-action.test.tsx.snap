// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`viewAction test suite test relax inject viewAction 1`] = `
<div>
  <div>
    
  </div>
  <div>
    
  </div>
</div>
`;

exports[`viewAction test suite test relax inject viewAction 2`] = `
<div>
  <div>
    hello
  </div>
  <div>
    world
  </div>
</div>
`;

exports[`viewAction test suite test relax inject viewAction 3`] = `
MockLog {
  "_logs": Array [
    "%cplume2@1.0.0ðŸš€",
    "StoreProvider(HelloApp) will mount ðŸš€",
    "Relax(HelloRelax) will mount rx store: true ðŸš€ ",
    "props:|>",
    "relaxProps:|>",
    "store dispatch => 'hello'",
    "params|> hello",
    "HelloWorldActor receive => 'hello'",
    "
StoreProvider(HelloApp) will update ðŸš€",
    "Relax(HelloRelax) will update rx store true ðŸš€",
    "props:|>",
    "relaxProps:|>",
    "store dispatch => 'world'",
    "params|> world",
    "HelloWorldActor receive => 'world'",
    "
StoreProvider(HelloApp) will update ðŸš€",
    "Relax(HelloRelax) will update rx store true ðŸš€",
    "props:|>",
    "relaxProps:|>",
  ],
}
`;
